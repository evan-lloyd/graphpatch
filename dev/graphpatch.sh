#!/bin/bash

function docker_binds() {
    echo "--mount type=bind,src=$(pwd)/graphpatch,dst=/graphpatch/graphpatch \
    --mount type=bind,src=$(pwd)/tests,dst=/graphpatch/tests \
    --mount type=bind,src=$(pwd)/demos,dst=/graphpatch/demos \
    --mount type=bind,src=$(pwd)/dev,dst=/graphpatch/dev \
    --mount type=bind,src=$(pwd)/pyproject.toml,dst=/graphpatch/pyproject.toml \
    --mount type=bind,src=$(pwd)/uv.lock,dst=/graphpatch/uv.lock \
    --mount type=bind,src=$(pwd)/README.md,dst=/graphpatch/README.md \
    --mount type=bind,src=$(pwd)/LICENSE,dst=/graphpatch/LICENSE \
    --mount type=bind,src=$(pwd)/.black,dst=/graphpatch/.black \
    --mount type=bind,src=$(pwd)/.flake8,dst=/graphpatch/.flake8 \
    --mount type=bind,src=$(pwd)/.isort.cfg,dst=/graphpatch/.isort.cfg \
    --mount type=bind,src=$(pwd)/.taplo.toml,dst=/graphpatch/.taplo.toml \
    --mount type=bind,src=$(pwd)/tox.ini,dst=/graphpatch/tox.ini \
    --mount type=bind,src=$(pwd)/pytest.ini,dst=/graphpatch/pytest.ini \
    --mount type=bind,src=$(pwd)/.python-version,dst=/graphpatch/.python-version"
}

function docker_run() {
    docker run --gpus all $(docker_binds) "$@" || docker run $(docker_binds) "$@"
}

function tailscale_env() {
    echo "-e TAILSCALE_AUTH_KEY=$TAILSCALE_AUTH_KEY -e TAILSCALE_HOST_NAME=graphpatch"
}
