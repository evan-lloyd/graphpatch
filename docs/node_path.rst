NodePath
##############
.. py:currentmodule:: graphpatch.meta

.. autoclass:: NodePath()
