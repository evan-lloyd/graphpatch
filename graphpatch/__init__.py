# Load bitsandbytes early so we can suppress its rather chatty startup process.
from .optional import bitsandbytes  # noqa: F401
from .patch import (
    AddPatch,
    CustomPatch,
    Patch,
    ProbePatch,
    RecordPatch,
    ReplacePatch,
    ZeroPatch,
)
from .patchable_graph import PatchableGraph

__all__ = [
    "AddPatch",
    "CustomPatch",
    "Patch",
    "ProbePatch",
    "RecordPatch",
    "ReplacePatch",
    "ZeroPatch",
    "PatchableGraph",
]
