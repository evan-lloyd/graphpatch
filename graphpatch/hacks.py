# mypy: ignore-errors

import operator
from contextlib import ExitStack, contextmanager
from functools import partial, partialmethod

import torch

TORCH_VERSION = tuple(int(v.split("+")[0]) for v in torch.__version__.split("."))


def fix_gpt2_bool_buffers(model):
    # Accelerate seems to convert bool buffers to float16 where it really shouldn't
    for name, buffer in model.named_buffers():
        if "attn.bias" not in name:
            continue
        split_name = name.split(".")
        module = model.get_submodule(".".join(split_name[:-1]))
        setattr(module, split_name[-1], buffer.to(torch.bool))


def clean_up_rotary_embedding(module, op):
    # TODO: figure out a more principled way to do this; particularly, why compile()
    # doesn't want to read seq_len from kwargs and instead bakes it in as a constant. Might
    # be specifically related to how slice operations are handled.
    getitem_node = next(
        n
        for n in module.graph.nodes
        if n.target == operator.getitem and n.args[0].name == f"{op}_cached"
    )
    seq_len_node = next(
        n for n in module.graph.nodes if n.op == "placeholder" and n.name == "seq_len"
    )
    with module.graph.inserting_after(seq_len_node):
        slice_node = module.graph.call_function(slice, (None, seq_len_node, None))
    getitem_node.args = getitem_node.args[0:1] + (
        getitem_node.args[1][:2] + (slice_node,) + getitem_node.args[1][3:],
    )


def patch_llama(graph_module, _original_module):
    """At the moment, LlamaRotaryEmbedding doesn't quite compile properly. It is supposed to read
    in the length of the current token sequence, but for some reason compile() converts the argument
    into a constant. This leads to run-time errors when trying to run the compiled model with
    different-length inputs.
    """
    # Improved dynamic shapes in 2.1 on means we don't have to do this.
    if TORCH_VERSION >= (2, 1):
        return
    for name, module in graph_module.named_modules():
        if "rotary_emb" in name:
            clean_up_rotary_embedding(module, "cos")
            clean_up_rotary_embedding(module, "sin")
            module.recompile()


@contextmanager
def monkeypatch_dynamic_shapes():
    """For torch >= 2.1.0. This version improves dynamic shapes in a way that's problematic for our
    use case; the philosophy seems to be one of eager optimization, with run-time checks that
    trigger a re-compile in case the optimistic assumptions are violated. Our current design doesn't
    work with that, since we assume a static GraphModule. In the future, we may want to refactor to
    roll with these re-compilations, in which case we can probably do away with this hack.
    """
    from torch._dynamo.output_graph import OutputGraph, SubgraphTracer
    from torch._dynamo.source import (
        DefaultsSource,
        LocalSource,
        NNModuleSource,
        TensorProperty,
        TensorPropertySource,
    )
    from torch._dynamo.utils import get_fake_value
    from torch._dynamo.variables import builder
    from torch._dynamo.variables.builder import VariableBuilder
    from torch.fx.experimental.symbolic_shapes import DimDynamic, ShapeEnv

    def wrap_literal(self, _original, value):
        # Avoids some additional cases of tensor sizes getting specialized.
        if type(value) is int and isinstance(self.get_source(), (LocalSource, NNModuleSource)):
            return self.wrap_unspecialized_primitive(value)
        return _original(self, value)

    def wrap_fx_proxy_cls(_original, target_cls, tx, *args, **kwargs):
        # Tweak some internal flags to avoid specializing on tensor dimensions.
        tx.fake_mode.shape_env.specialize_zero_one = False
        tx.fake_mode.shape_env.duck_shape = False

        # Tensor sizes of 1 keep getting specialized via some mysterious codepath. We can prevent
        # this by overriding the outputs of any nodes retrieving tensor sizes with only symbolic
        # sizes.
        proxy = kwargs.get("proxy")
        if (
            proxy is not None
            and proxy.node
            and proxy.node.op == "call_method"
            and proxy.node.target == "size"
        ):
            fake_value = get_fake_value(proxy.node, tx)

            def enforce_symbolic(t):
                i, v = t
                if isinstance(v, int):
                    source = TensorPropertySource(
                        LocalSource(local_name=proxy.node.name), TensorProperty.SIZE, i
                    )
                    return tx.fake_mode.shape_env.create_symintnode(
                        tx.fake_mode.shape_env.create_symbol(
                            v,
                            source,
                            dynamic_dim=DimDynamic.DYNAMIC,
                        ),
                        hint=v,
                        source=source,
                    )
                return v

            if isinstance(fake_value, torch.Size):
                fake_value = torch.Size(map(enforce_symbolic, enumerate(fake_value)))
                kwargs["example_value"] = fake_value

        return _original(target_cls, tx, *args, **kwargs)

    def produce_guards(self, _original, *args, **kwargs):
        return []

    def _maybe_guard_eq(self, _original, *args, **kwargs):
        """This prevents many cases of torch deciding to specialize on tensor dimensions. We don't
        care about the guards that would have gotten generated because they aren't present in the
        compiled GraphModule, and we discard the OptimizedModule after compiling."""
        return

    def remove_unused_graphargs(self, _original):
        """Remove the dynamic size placeholders, since we monkeypatched away any logic that uses
        them.
        """
        # NB: *not* calling original; we only want to remove shape guard SymInts, since we
        # monkeypatched installing the related guards. Also cached default values.
        for node in self.placeholders:
            arg = node.meta["grapharg"]
            if isinstance(arg.source, (TensorPropertySource, DefaultsSource)):
                self.remove_node(node)

    def create_graph_input(self, _original, name, *args, **kwargs):
        """Unmangles input names, which end up with something hideous like L__foo_"""
        source = kwargs.get("source")
        if source is not None:
            if isinstance(source, NNModuleSource) and name.startswith("L_self_"):
                name = name.replace("L_self_", "", 1)
            elif hasattr(source, "local_name"):
                name = source.local_name
        return _original(self, name, *args, **kwargs)

    patch_map = {
        SubgraphTracer: [create_graph_input],
        OutputGraph: [remove_unused_graphargs],
        ShapeEnv: [_maybe_guard_eq, produce_guards],
        builder: [wrap_fx_proxy_cls],
        VariableBuilder: [wrap_literal],
    }
    orig_functions = {
        patched_obj: {a.__name__: getattr(patched_obj, a.__name__) for a in attrs}
        for patched_obj, attrs in patch_map.items()
    }

    try:
        for patched_obj, attr_map in orig_functions.items():
            for attr, fn in attr_map.items():
                if isinstance(patched_obj, type):
                    setattr(patched_obj, attr, partialmethod(locals()[attr], fn))
                else:
                    setattr(patched_obj, attr, partial(locals()[attr], fn))
        yield
    finally:
        for patched_obj, attr_map in orig_functions.items():
            for attr, fn in attr_map.items():
                setattr(patched_obj, attr, fn)


@contextmanager
def monkeypatch_graph_names():
    """Monkeypatches OutputGraph.register_attr_or_module, which was the minimum intervention I
    found to reliably override the "self_" prefix that compile() adds to getattr and submodule node
    names in the compiled graph. We can likely do this in a less hacky way with a custom
    backend/tracer.
    """
    from torch._dynamo.output_graph import OutputGraph

    orig_method = OutputGraph.register_attr_or_module

    def strip_self_from_names(*args, **kwargs):
        [self, target, *names] = args

        def replace(name):
            if not isinstance(name, str):
                return name
            if name.startswith("self."):
                name = name.replace("self.", "", 1)
            if name.startswith("L['self']."):
                name = name.replace("L['self'].", "", 1)
            return name

        names = list(map(replace, names))
        return orig_method(self, target, *names, **kwargs)

    try:
        OutputGraph.register_attr_or_module = strip_self_from_names
        yield
    finally:
        OutputGraph.register_attr_or_module = orig_method


@contextmanager
def make_dynamo_ignore_hooks():
    """Only use for torch >= 2.1.0. In that version, torch attempts to compile any hooks you have
    applied to your model, but we need it to ignore them (as it did in 2.0.*) so we can easily
    record module inputs/outputs without affecting the compiled code.
    """
    from torch._dynamo import output_graph
    from torch._dynamo.utils import nnmodule_has_hooks as orig_function
    from torch._dynamo.variables import nn_module

    modules_to_patch = [nn_module, output_graph]

    def dummy_has_hooks(*args, **kwargs):
        return False

    try:
        for m in modules_to_patch:
            m.nnmodule_has_hooks = dummy_has_hooks
        yield
    finally:
        for m in modules_to_patch:
            m.nnmodule_has_hooks = orig_function


def get_size(target, index):
    return target[index]


def maybe_replace_dynamo_get_item_lambda(node):
    """In compiling GPT2-XL, torch.compile() leaves a call to a local function which makes it
    unpicklable. All it does is retrieve the size of the given torch.Size at the given index:
    https://github.com/pytorch/pytorch/blob/e9ebda29d87ce0916ab08c06ab26fd3766a870e5/torch/_dynamo/variables/lists.py#L391
    We can just do that in a function accessible from global scope, and become picklable again.
    """
    if getattr(node.target, "__name__", None) == "_dynamo_get_item_lambda":
        node.target = get_size
        if node.meta:
            node.meta["source_fn"] = get_size


@contextmanager
def set_dynamo_config():
    """Reconfigure some dynamo options for >= 2.1.0 to get compilations we can work with."""
    config_values = {
        "specialize_int": False,
        "assume_static_by_default": False,
        "automatic_dynamic_shapes": False,
        "capture_scalar_outputs": True,
        "capture_dynamic_output_shape_ops": True,
    }
    orig_values = {key: getattr(torch._dynamo.config, key) for key in config_values}
    for key, value in config_values.items():
        setattr(torch._dynamo.config, key, value)
    try:
        yield
    finally:
        for key, value in orig_values.items():
            setattr(torch._dynamo.config, key, value)


def dynamo_hacks_for_current_torch_version():
    hack_stack = ExitStack()
    if TORCH_VERSION >= (2, 1):
        hack_stack.enter_context(set_dynamo_config())
        hack_stack.enter_context(make_dynamo_ignore_hooks())
        hack_stack.enter_context(monkeypatch_dynamic_shapes())
    hack_stack.enter_context(monkeypatch_graph_names())
    return hack_stack
