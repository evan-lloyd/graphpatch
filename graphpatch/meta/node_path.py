from pprint import pformat
from typing import Any, ClassVar, List, Optional, Union, cast

from ..optional.dataclasses import dataclass
from . import GraphMeta, NodeData, NodeMeta, NodeShape, PrettyPrintedNodeData
from .node_data import MaybeNodeDataType


@dataclass(kw_only=True, repr=False)
class NodePath(PrettyPrintedNodeData[Union[GraphMeta, NodeMeta]]):
    """Helper class to set up autocomplete for finding nodes in notebooks and REPL."""

    MAX_COMPLETIONS: ClassVar[int] = 20
    _node_meta: Optional[NodeMeta] = None

    def __hash__(self) -> int:
        return self._path.__hash__()

    def __str__(self) -> str:
        return self.to_path()

    def __dir__(self) -> List[str]:
        return super().__dir__() + ["_code", "_shape"]

    def __getattr__(self, key: str) -> Any:
        if key == "_code":
            if isinstance(self._value, (GraphMeta, NodeMeta)):
                return self._value.code
            elif isinstance(self._node_meta, NodeMeta):
                return self._node_meta.code
        if key == "_shape":
            if self._children is not NodeData._NO_VALUE:
                return self.map(
                    lambda _, v: v,
                    lambda **kwargs: NodeShapePath(
                        show_containers=True,
                        unwrap_leaves=lambda value: value._shape,
                        **kwargs,
                    ),
                )
            if isinstance(self._value, NodeShape):
                return self._value._shape
            elif isinstance(self._value, NodeMeta):
                return self._value.shape
        return super().__getattr__(key)

    def to_path(self, allow_internal: bool = True) -> str:
        if not allow_internal and (
            self._value is NodeData._NO_VALUE or self._children is not NodeData._NO_VALUE
        ):
            completions = [n._path for _, n in self._iter_items() if n.is_leaf]
            if len(completions) > NodePath.MAX_COMPLETIONS:
                completions = completions[: NodePath.MAX_COMPLETIONS] + ["..."]
            raise ValueError(
                f"Incomplete node path: '{self._value}'. Possible completions:\n"
                + pformat(
                    completions,
                    indent=4,
                    compact=True,
                )
            )
        return self._path


class NodeShapePath(PrettyPrintedNodeData[NodeShape]):
    """Separate wrapper for NodeShape, which doesn't have custom handling for _code and _shape."""

    def __dir__(self) -> List[str]:
        return PrettyPrintedNodeData.__dir__(self)

    def __getattr__(self, key: str) -> Any:
        return PrettyPrintedNodeData.__getattr__(self, key)


def wrap_node_path(meta: NodeData[Union[GraphMeta, NodeMeta]]) -> NodePath:
    def make_node_path(**kwargs: Any) -> NodePath:
        node_path = kwargs["_path"]
        node = kwargs["_value"]

        def shape_value(
            path: str, value: MaybeNodeDataType[NodeShape]
        ) -> MaybeNodeDataType[NodeShape]:
            return value

        def make_shape(**kwargs: Any) -> NodePath:
            shape_path = kwargs["_path"]
            kwargs["_path"] = f"{node_path}|{shape_path}"
            return NodePath(show_containers=True, _node_meta=node, **kwargs)

        if (
            isinstance(node, NodeMeta)
            and kwargs["_children"] is NodeData._NO_VALUE
            and node.shape is not None
        ):
            kwargs["_original_type"] = node.shape._original_type
            kwargs["_value"] = node.shape._value
            kwargs["_node_meta"] = node
            kwargs["_children"] = cast(NodePath, node.shape.map(shape_value, make_shape))._children

        return NodePath(show_containers=True, **kwargs)

    def maybe_meta(
        path: str,
        value: MaybeNodeDataType[Union[GraphMeta, NodeMeta]],
    ) -> MaybeNodeDataType[Union[GraphMeta, NodeMeta]]:
        return value

    node_path = cast(
        NodePath,
        meta.map(
            maybe_meta,
            make_node_path,
        ),
    )
    return node_path
