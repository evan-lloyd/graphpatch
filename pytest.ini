[pytest]
testpaths =
    tests
filterwarnings = 
    ignore::UserWarning:bitsandbytes
    ignore:Overriding torch_dtype=None
    ignore:invalid escape sequence:DeprecationWarning
    ignore:`_is_quantized_training_enabled` is going to be deprecated:FutureWarning
