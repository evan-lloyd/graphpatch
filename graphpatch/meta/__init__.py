from .graph_meta import (
    GraphMeta,
    NodeMeta,
    NodeShape,
    wrap_graph_module,
    wrap_node_shape,
    wrap_output_argument_index,
)
from .node_data import (
    NodeData,
    PrettyPrintedNodeData,
    make_pretty_printed,
    wrap_node_data,
)
from .node_path import NodePath, wrap_node_path

__all__ = [
    "GraphMeta",
    "NodeData",
    "PrettyPrintedNodeData",
    "NodeMeta",
    "NodePath",
    "NodeShape",
    "make_pretty_printed",
    "wrap_output_argument_index",
    "wrap_node_data",
    "wrap_graph_module",
    "wrap_node_path",
    "wrap_node_shape",
]
