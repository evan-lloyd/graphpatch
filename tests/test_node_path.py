from typing import Optional

import pytest
from syrupy.extensions.json import JSONSnapshotExtension
from syrupy.extensions.amber.serializer import AmberDataSerializer

from graphpatch.hacks import TORCH_VERSION
from graphpatch.meta import NodeData, NodeMeta, wrap_node_path
from graphpatch.meta.graph_meta import WrappedCode
from graphpatch.meta.node_path import NodeShapePath
from tests.fixtures.deeply_nested_output_module import DeeplyNestedOutputModule


class SingleFileAmber(JSONSnapshotExtension):
    _file_extension = "ambr"

    def serialize(self, data, **kwargs):
        return AmberDataSerializer.serialize(data, **kwargs)

    @classmethod
    def dirname(cls, *, test_location):
        original = JSONSnapshotExtension.dirname(test_location=test_location)
        return f"{original}/{test_location.testname}"

    @classmethod
    def get_snapshot_name(cls, *, test_location, index):
        return str(index)


class MockNodeMeta(NodeMeta):
    shape: Optional[NodeData[int]]
    hidden: bool = False

    def __init__(self, shape):
        self.shape = shape
        self.hidden = False


@pytest.fixture
def path_data():
    return NodeData(
        _path="",
        _children={
            "foo": NodeData(
                _path="foo",
                _children={
                    "bar": NodeData(
                        _path="foo.bar",
                        _children={
                            "sub_0": NodeData(
                                _path="foo.bar.sub_0",
                                _value=MockNodeMeta(
                                    shape=NodeData(
                                        _path="",
                                        _children={
                                            "sub_0": NodeData(
                                                _path="sub_0",
                                                _children={
                                                    "baz": NodeData(
                                                        _path="sub_0.baz",
                                                        _value=0,
                                                        _original_type=int,
                                                    )
                                                },
                                                _original_type=dict,
                                            ),
                                            "sub_1": NodeData(
                                                _path="sub_1",
                                                _value=0,
                                                _original_type=int,
                                            ),
                                        },
                                        _original_type=tuple,
                                    )
                                ),
                                _original_type=dict,
                            ),
                            "sub_1": NodeData(
                                _path="foo.bar.sub_1",
                                _value=MockNodeMeta(shape=None),
                                _original_type=tuple,
                            ),
                        },
                        _original_type=tuple,
                    ),
                },
                _original_type=dict,
            ),
        },
        _original_type=dict,
    )


def test_wrap_node_path(path_data):
    node_path = wrap_node_path(path_data)

    # Node hierarchy should have been preserved
    for key in [
        "",
        "foo",
        "foo.bar",
        "foo.bar.sub_0",
        "foo.bar.sub_1",
    ]:
        assert node_path._dig(key)._path == key

    # Append shape paths following "|"
    for key in ["sub_0", "sub_0.baz", "sub_1"]:
        assert node_path._dig(f"foo.bar.sub_0.{key}")._path == f"foo.bar.sub_0|{key}"

    # We should be able to access each path by getattr
    assert node_path.foo is node_path._dig("foo")
    assert node_path.foo.bar is node_path._dig("foo.bar")
    assert node_path.foo.bar.sub_0 is node_path._dig("foo.bar.sub_0")
    assert node_path.foo.bar.sub_1 is node_path._dig("foo.bar.sub_1")
    assert node_path.foo.bar.sub_0.sub_0 is node_path._dig("foo.bar.sub_0.sub_0")
    assert node_path.foo.bar.sub_0.sub_0.baz is node_path._dig("foo.bar.sub_0.sub_0.baz")
    assert node_path.foo.bar.sub_0.sub_1 is node_path._dig("foo.bar.sub_0.sub_1")


def test_node_path_autocomplete(path_data):
    node_path = wrap_node_path(path_data)

    assert set(node_path.__dir__()) == {"foo", "_code", "_shape"}
    assert set(node_path.foo.__dir__()) == {"bar", "_code", "_shape"}
    assert set(node_path.foo.bar.__dir__()) == {"sub_0", "sub_1", "_code", "_shape"}
    assert set(node_path.foo.bar.sub_0.__dir__()) == {"sub_0", "sub_1", "_code", "_shape"}
    assert set(node_path.foo.bar.sub_0.sub_0.__dir__()) == {"baz", "_code", "_shape"}
    assert set(node_path.foo.bar.sub_0.sub_0.baz.__dir__()) == {"_code", "_shape"}
    assert set(node_path.foo.bar.sub_0.sub_1.__dir__()) == {"_code", "_shape"}


@pytest.mark.parametrize("all_patchable_graphs", ["compiled", "opaque"], indirect=True)
def test_patchable_graph_graph_repr(all_patchable_graphs, snapshot):
    snapshot = snapshot.with_defaults(extension_class=SingleFileAmber)

    # Note we get slightly different structures for newer versions of torch, due to it retaining
    # more not-actually-used nodes. In future releases we should use our own logic to clean up
    # graphs after extraction, which should eliminate this discrepancy.
    if TORCH_VERSION >= (2, 4):
        torch_version_suffix = "2_4"
    elif TORCH_VERSION >= (2, 2):
        torch_version_suffix = "2_2-2_3"
    elif TORCH_VERSION >= (2, 1):
        torch_version_suffix = "2_1"
    else:
        torch_version_suffix = "2_0"

    for pg_name, pg in all_patchable_graphs.items():
        assert repr(pg.graph) == snapshot(
            name=f"{pg_name}_{torch_version_suffix}"
        ), f"Snapshot mismatch for {pg_name}"


def test_node_path_shape(patchable_deeply_nested_output_module):
    pg = patchable_deeply_nested_output_module
    # We should be able to throw a "._shape" in anywhere
    assert pg.graph._shape.add == DeeplyNestedOutputModule._shape
    assert pg.graph.add._shape == DeeplyNestedOutputModule._shape
    assert pg.graph.output.sub_0.sub_0._shape == DeeplyNestedOutputModule._shape
    assert pg.graph.output.sub_0._shape.sub_0 == DeeplyNestedOutputModule._shape
    assert pg.graph.output._shape.sub_0.sub_0 == DeeplyNestedOutputModule._shape
    assert pg.graph._shape.output.sub_0.sub_0 == DeeplyNestedOutputModule._shape


def test_node_path_code(patchable_deeply_nested_output_module, snapshot):
    pg = patchable_deeply_nested_output_module
    if TORCH_VERSION >= (2, 1):
        assert pg.graph._code == snapshot(name="2_1")
    else:
        assert pg.graph._code == snapshot(name="2_0")
    assert str(pg.graph.output._code) == (
        "return ((linear_0,), [([getitem_5], getitem_6), ([getitem_9], getitem_10), ([getitem_13], getitem_14)],"
        " {'nested_dict': [(linear_1,)]})"
    )


def test_protected_names(patchable_protected_name_module, protected_name_module_inputs):
    pg = patchable_protected_name_module
    pg(protected_name_module_inputs)

    # Special REPL properties should not have been shadowed by nodes.
    assert isinstance(pg.graph._code, WrappedCode)
    assert isinstance(pg.graph._shape, NodeShapePath)
    assert "_code" not in pg.graph
    assert "_shape" not in pg.graph

    # We should *not* have changed the targets of any placeholders, since that would break
    # calling code if using kwargs.
    base_nodes = list(pg._graph_module.graph.nodes)
    assert [n.target for n in base_nodes if n.op == "placeholder"] == [
        "_shape",
        "_code",
        "sub_shape",
        "_graphpatch_args",
    ]

    # The renamed node should still be in the graph.
    assert any(n.node.name == "_code" for n in pg._meta.values() if n.node is not None)
