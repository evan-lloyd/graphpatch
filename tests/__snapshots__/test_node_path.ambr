# serializer version: 1
# name: test_node_path_code[2_0]
  def forward(self, x : torch.Tensor):
      child_a = self.child_a(x);  x = None
      getitem = child_a[0]
      getitem_1 = getitem[0];  getitem = None
      getitem_2 = child_a[1];  child_a = None
      getitem_3 = getitem_2[0]
      getitem_4 = getitem_3[0]
      getitem_5 = getitem_4[0];  getitem_4 = None
      getitem_6 = getitem_3[1];  getitem_3 = None
      getitem_7 = getitem_2[1]
      getitem_8 = getitem_7[0]
      getitem_9 = getitem_8[0];  getitem_8 = None
      getitem_10 = getitem_7[1];  getitem_7 = None
      getitem_11 = getitem_2[2];  getitem_2 = None
      getitem_12 = getitem_11[0]
      getitem_13 = getitem_12[0];  getitem_12 = None
      getitem_14 = getitem_11[1];  getitem_11 = None
      linear_0 = getattr(self.linear, "0")(getitem_1);  getitem_1 = None
      add = linear_0 + 2
      linear_1 = getattr(self.linear, "1")(add);  add = None
      return ((linear_0,), [([getitem_5], getitem_6), ([getitem_9], getitem_10), ([getitem_13], getitem_14)], {'nested_dict': [(linear_1,)]})
# ---
# name: test_node_path_code[2_1]
  def forward(self, x : torch.Tensor):
      child_a = self.child_a(x);  x = None
      getitem = child_a[0]
      getitem_1 = getitem[0];  getitem = None
      getitem_2 = child_a[1];  child_a = None
      getitem_3 = getitem_2[0]
      getitem_4 = getitem_3[0]
      getitem_5 = getitem_4[0];  getitem_4 = None
      getitem_6 = getitem_3[1];  getitem_3 = None
      getitem_7 = getitem_2[1]
      getitem_8 = getitem_7[0]
      getitem_9 = getitem_8[0];  getitem_8 = None
      getitem_10 = getitem_7[1];  getitem_7 = None
      getitem_11 = getitem_2[2];  getitem_2 = None
      getitem_12 = getitem_11[0]
      getitem_13 = getitem_12[0];  getitem_12 = None
      getitem_14 = getitem_11[1];  getitem_11 = None
      linear_0 = getattr(self.linear, "0")(getitem_1);  getitem_1 = None
      add = linear_0 + 2
      linear_1 = getattr(self.linear, "1")(add);  add = None
      return ((linear_0,), [([getitem_5], getitem_6), ([getitem_9], getitem_10), ([getitem_13], getitem_14)], {'nested_dict': [(linear_1,)]})
# ---
